cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/discovery/KubeDiscoveryService.java

```diff
@@ -266,15 +266,24 @@ private boolean createV1Service(CoreV1Api api, String serviceName, Discoverable
 
     service.setSpec(spec);
 
-    try {
-      api.createNamespacedService(namespace, service, null, null, null);
-      LOG.info("Service created in kubernetes with name {} and port {}", serviceName, port.getPort());
-    } catch (ApiException e) {
-      // It means the service already exists. In this case we update the port if it is not the same.
-      if (e.getCode() == HttpURLConnection.HTTP_CONFLICT) {
-        return false;
+    boolean serviceCreated = false;
+    int maxAttempts = 3;
+
+    for (int i = 1; i <= maxAttempts && !serviceCreated; ++i) {
```

Comment:
I still think we shouldn't retry in the discovery service. If any form of retries is needed, it should be from the caller. However, I think it is sufficient that on service registration failure, we make sure the main process gets terminated and let K8s to restart the pod.

Reply:
> I still think we shouldn't retry in the discovery service. If any form of retries is needed, it should be from the caller.

Sounds good. The other context that seemed relevant is in the '[service start](https://gist.github.com/domnicjp/949adcaecdd84a528f6e89c941108205)' - not sure if that is worthwhile pursuing.

> However, I think it is sufficient that on service registration failure, we make sure the main process gets terminated and let K8s to restart the pod.

As noted [above](https://github.com/cdapio/cdap/pull/12443#issuecomment-665851976) - the tests that I performed for intentional errors(or rather intentional exceptions thrown) in both the `create`/'service registration' and `update` codepaths seemed to suggest repeated retries to start the `appfabric` pod i.e. I do not see a 'wedged state' wherein the 'main process' remains 'un-terminated' or is preventing `k8s` from attempting to restart the pod.

Let me know your thoughts. (And thanks for your help with this!)